ocs_ci/utility/azure_utils.py

```diff
@@ -71,23 +93,81 @@ def get_node_by_attached_volume(self, volume):
 
         for vm in vm_list:
             for disk in vm.storage_profile.data_disks:
-                if disk.name == volume:
+                if disk.name == volume.name:
                     return vm
 
-    def detach_volume(self, volume, vm, timeout=120):
-        pass
+    def detach_volume(self, volume, node, timeout=120):
+        """
+        Detach volume if attached
 
+        Args:
+            volume (disk): disk object required to delete a volume
+            node (OCS): The OCS object representing the node
+            timeout (int): Timeout in seconds for API calls
 
-def get_data_volumes(deviceset_pvs):
-    """
-    Get the instance data volume names
+        """
+        vm = self.get_vm_instance(node.name)
+        data_disks = vm.storage_profile.data_disks
+        data_disks[:] = [disk for disk in data_disks if disk.name != volume.name]
+        logger.info(
+            "Detaching volume: %s Instance: %s", volume.name, vm.name
+        )
+        result = self.compute_client.virtual_machines.create_or_update(
+            self._resourcegroup,
+            vm.name,
+            vm)
+        result.wait()
+        for x in range(timeout):
+            disk_state = self.get_disk_state(volume.name)
+            logger.debug(
+                "Volume id: %s has status: %s", volume.name,
+                disk_state
+            )
+            if disk_state == 'Unattached':
+                break
+            if x == timeout - 1:
```

Comment:
I would rather use more forgiving check, like `x >= timeout - 1`, so that if last get disk state operation takes too long, we don't end up in an endless loop.

Reply:
`timeout` is the range value in the for loop.

- x iterates from 0 to `timeout-1` and x exceeding `timeout-1` is not possible
- At the end of each iteration `time.sleep(1)` is used to sleep for one second.

Reply:
please consider using ocs_ci.utility.utils.TimeoutSampler

Reply:
ah, ok - you are just checking that the loop is in it's last iteration

Reply:
@ebenahar Thanks, I have replaced with TimeoutSampler wherever possible.